### PR TITLE
Add TaskBuilder (2018 attempt)

### LINKED ISSUE #49: Builder interface
By now, our library interface is ugly as hell (see #47 - way too much `None` passing to the constructor, not very self-explanatory).

We should add a `TaskBuilder` where one can do

```rust
TaskBuilder::new()
    .with_description("Foo")
    .with_date_parsed("2017-01-01 12:13:14", "%Y-%m-%d %H:%M:%S")
    .status(TaskStatus::Pending)
    .with_annotation("Foo".to_string())
    .with_annotation("Bar".to_string())
    .depends_on(some_uuid)
    .build()
```

or something like this. Maybe we can play with `build_derive` or whatever the crate name was of this nice builder helper crate.

Probably another release then...

### PR DESCRIPTION
**Pull request by maralorn**
I took your PR and fixed it, until it worked.
Includes changelog entry and tests.
The Builder does not have its own namespace (as in task-hookrs::builder:: or something) I didn't know how to do that and figured it's not that important.
Closes #49

### CODE REVIEW

**Comment by matthiasbeyer** [comment]
Hmh, I'd like to have a namespace for it (`::builder::TaskBuilder` seems fine)

I'm on mobile, but what I saw so far looks good.
Make sure to squash the fixup commits before merging (I guess we should block such branches via CI... Maybe that's a good idea. Imag has a setup for this, if you care enough ... :smile:)

**Comment by maralorn** [comment]
Do I have to squash them manually or can I do some autosquash magic?
I tried and failed with the autosquash …

**Comment by maralorn** [comment]
Hm, I don't think that derive_builder has that ability and I don't know how to work around it.
We can of course re-export in ::builder, but I don't know how to hide it in ::. Is it perhaps not pub anyway?

File: src/task.rs [hunk 1]
```diff
@@ -847,4 +893,76 @@ mod test {
         assert!(back.contains("1234"));
     }
 
+    #[test]
+    fn test_builder_simple() {
+        use task::TaskBuilder;
+
+        let t = TaskBuilder::default()
+            .description("test".to_owned())
```
Comment (matthiasbeyer) [inline src/task.rs]: Do we need the `.to_owned()` part if the builder is a `into` builder?

File: src/task.rs [hunk 2]
```diff
@@ -847,4 +893,76 @@ mod test {
...
+            .entry(mkdate("20150619T165438Z"))
+            .build();
+        println!("{:?}", t);
+        assert!(t.is_ok());
+        let t = t.unwrap();
+
+        assert_eq!(t.status(), &TaskStatus::Pending);
+        assert_eq!(t.description(), &"test".to_owned());
```
Comment (matthiasbeyer) [inline src/task.rs]: Same here.

File: src/task.rs (hunk 1 above)
Comment (maralorn) [reply src/task.rs]: That's true. I wrote that test before introducing the `into` builder.

File: src/task.rs (hunk 2 above)
Comment (maralorn) [reply src/task.rs]: Well I found more than one place, were removing `.to_owned()` was possible. Didn't know, that `&str´ implements partialeq for string.

**Comment by maralorn** [comment]
I rebased onto master and fixed the required changes.

**Pull request merged**